sixtyfps_runtime/corelib/graphics/fps_counter.rs

```diff
@@ -0,0 +1,70 @@
+/* LICENSE BEGIN
+    This file is part of the SixtyFPS Project -- https://sixtyfps.io
+    Copyright (c) 2021 Olivier Goffart <olivier.goffart@sixtyfps.io>
+    Copyright (c) 2021 Simon Hausmann <simon.hausmann@sixtyfps.io>
+
+    SPDX-License-Identifier: GPL-3.0-only
+    This file is also available under commercial licensing terms.
+    Please contact info@sixtyfps.io for more information.
+LICENSE END */
+/*!
+This module contains a simple helper type to measure the average number of frames rendered per second.
+*/
+
+use crate::timers::*;
+use std::cell::RefCell;
+use std::rc::Rc;
+
+/// Helper class that rendering backends can use to provide an FPS counter
+pub struct FPSCounter {
+    frame_times: RefCell<Vec<instant::Instant>>,
+    update_timer: Timer,
+}
+
+impl FPSCounter {
+    /// Returns a new instance of the counter if requested by the user (via `SIXTYFPS_DEBUG_PERFORMANCE` environment variable).
+    pub fn new() -> Option<Rc<Self>> {
+        std::env::var("SIXTYFPS_DEBUG_PERFORMANCE").is_ok().then(|| {
+            Rc::new(Self { frame_times: Default::default(), update_timer: Default::default() })
+        })
+    }
+
+    /// Call this function if you want to start measurements. This will also print out some system information such as whether
+    /// this is a debug or release build, as well as the provided winsys_info string.
+    pub fn start(self: &Rc<Self>, winsys_info: &str) {
+        #[cfg(debug_assertions)]
+        let build_config = "debug";
+        #[cfg(not(debug_assertions))]
+        let build_config = "release";
+
+        eprintln!(
+            "sixtyfps: starting fps measurements. build config: {}; backend/windowing system: {}",
+            build_config, winsys_info
+        );
+
+        let this = self.clone();
+        self.update_timer.stop();
+        self.update_timer.start(TimerMode::Repeated, std::time::Duration::from_secs(1), move || {
+            this.refresh();
+        })
+    }
+
+    /// Call this function every time you've completed the rendering of a frame.
+    pub fn measure_frame_rendered(self: &Rc<Self>) {
+        self.frame_times.borrow_mut().push(instant::Instant::now());
+    }
+
+    fn refresh(self: &Rc<Self>) {
+        let mut i = 0;
+        let mut frame_times = self.frame_times.borrow_mut();
+        while i < frame_times.len() {
+            if frame_times[i].elapsed() > std::time::Duration::from_secs(1) {
+                frame_times.remove(i);
+            } else {
+                i += 1
+            }
+        }
+
+        eprintln!("average frames per second: {}", frame_times.len());
```

Comment:
I'd like to use `debug_log!` here, but I can't seem to figure out how.

Just using `debug_log!` I get this:

```
error: cannot find macro `debug_log` in this scope
  --> sixtyfps_runtime/corelib/graphics/fps_counter.rs:68:9
   |
68 |         debug_log!("average frames per second: {}", frame_times.len());
   |         ^^^^^^^^^
   |
   = note: consider importing this macro:
           crate::debug_log
   = help: have you added the `#[macro_use]` on the module/import?
```

If I follow the "advice" and write `use crate::debug_log;` in the line before, I get this:

```
error: cannot determine resolution for the import
  --> sixtyfps_runtime/corelib/graphics/fps_counter.rs:68:13
   |
68 |         use crate::debug_log;
   |             ^^^^^^^^^^^^^^^^
   |
   = note: import resolution is stuck, try simplifying other imports

error: cannot determine resolution for the macro `debug_log`
  --> sixtyfps_runtime/corelib/graphics/fps_counter.rs:69:9
   |
69 |         debug_log!("average frames per second: {}", frame_times.len());
   |         ^^^^^^^^^
   |
   = note: import resolution is stuck, try simplifying macro imports

error: macro-expanded `macro_export` macros from the current crate cannot be referred to by absolute paths
   --> sixtyfps_runtime/corelib/graphics/fps_counter.rs:68:13
    |
68  |         use crate::debug_log;
    |             ^^^^^^^^^^^^^^^^
    |
    = note: `#[deny(macro_expanded_macro_exports_accessed_by_absolute_paths)]` on by default
    = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
    = note: for more information, see issue #52234 <https://github.com/rust-lang/rust/issues/52234>
note: the macro is defined here
   --> sixtyfps_runtime/corelib/tests.rs:111:9
    |
111 | /         macro_rules! debug_log {
112 | |             ($($t:tt)*) => (eprintln!($($t)*))
113 | |         }
    | |_________^

error: could not compile `sixtyfps-corelib` due to 3 previous errors
```

So instead of `crate` if I use `use super::super::tests::debug_log;` I get:

```
error[E0432]: unresolved import `super::super::tests::debug_log`
  --> sixtyfps_runtime/corelib/graphics/fps_counter.rs:68:13
   |
68 |         use super::super::tests::debug_log;
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `debug_log` in `tests`
   |
   = note: this could be because a macro annotated with `#[macro_export]` will be exported at the root of the crate instead of the module where it is defined
help: a macro with this name exists at the root of the crate
   |
68 |         use ::debug_log;
   |             ~~~~~~~~~~~

error: cannot determine resolution for the macro `debug_log`
  --> sixtyfps_runtime/corelib/graphics/fps_counter.rs:69:9
   |
69 |         debug_log!("average frames per second: {}", frame_times.len());
   |         ^^^^^^^^^
   |
   = note: import resolution is stuck, try simplifying macro imports
```

And if I then use `use ::debug_log;` I get this:

```
error[E0432]: unresolved import `debug_log`
  --> sixtyfps_runtime/corelib/graphics/fps_counter.rs:68:13
   |
68 |         use ::debug_log;
   |             ^^^^^^^^^^^ no external crate `debug_log`

error: cannot determine resolution for the macro `debug_log`
  --> sixtyfps_runtime/corelib/graphics/fps_counter.rs:69:9
   |
69 |         debug_log!("average frames per second: {}", frame_times.len());
   |         ^^^^^^^^^
   |
   = note: import resolution is stuck, try simplifying macro imports
```

Finally `use super::super::debug_log;` also doesn't work.

Do you have any idea what I may be missing here?

Reply:
Macro needs to be declared before their used.  So make  sure that you declare the module that has this macro before the one that uses it.

Reply:
Ahh, thank you! that works

Reply:
Ah no, that also doesn't help. I thought it did, but my build was a "red herring" :)